test/distributed/checkpoint/e2e/test_fsdp.py

```diff
@@ -0,0 +1,96 @@
+# Owner(s): ["oncall: distributed"]
+import torch
+import torch.distributed as dist
+import torch.distributed.checkpoint as DCP
+import torch.nn as nn
+from torch.distributed._tensor.device_mesh import init_device_mesh
+from torch.distributed.checkpoint.state_dict import get_state_dict, set_state_dict
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
+    parametrize,
+    run_tests,
+)
+
+from torch.testing._internal.distributed._tensor.common_dtensor import (
+    DTensorTestBase,
+    skip_if_lt_x_gpu,
+    with_comms,
+)
+from torch.testing._internal.distributed.checkpoint_utils import with_temp_dir
+
+# Simple and boring model to test interface and some corner cases that do not
+# require complicated wrapping strategy.
+class TestDummyModel(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        # torch.manual_seed(0)
+        self.net1 = nn.Sequential(nn.Linear(8, 16), nn.ReLU())
+        self.net2 = nn.Sequential(nn.Linear(16, 32), nn.ReLU())
+        self.net3 = nn.Linear(32, 64)
+        self.net4 = nn.Sequential(nn.ReLU(), nn.Linear(64, 8))
+
+    def forward(self, x):
+        return self.net4(self.net3(self.net2(self.net1(x))))
+
+    def get_input(self):
+        return torch.rand(8, 8, device="cuda")
+
+
+class TestFSDP(DTensorTestBase):
+    def _create_model(self, device_mesh=None):
+        dummy_model = TestDummyModel()
+
+        model = FSDP(dummy_model.cuda(), device_mesh=device_mesh, use_orig_params=True)
+        optim = torch.optim.Adam(model.parameters(), lr=0.1)
+        model(model.get_input()).sum().backward()
+        optim.step()
+
+        return model, optim
+
+    def _equal_state_dict(self, model_0, model_1):
```

Comment:
@fegin Is your recommendation (https://github.com/pytorch/pytorch/blob/main/test/distributed/tensor/parallel/test_fsdp_2d_parallel.py#L638-L641) necessary here, or do you think this is sufficient?

Reply:
Assuming we'll need something similar to compare optimizers?

Reply:
Chien-Chin has some utils to compare state_dict here. See if anything is useful for you, and maybe we can just create a test utils for these. https://github.com/pytorch/pytorch/blob/main/test/distributed/checkpoint/test_state_dict.py#L56

Reply:
I believe since I can compare DTensor's directly with `torch.equal` we don't really need this logic, so I left the model state dict comparison the same for now.

I also opted for a direct comparison of the optim state dict, since it was a bit simpler and that seems to work here as well - https://github.com/pytorch/pytorch/blob/5a6f8014c4c2480f77d40fa214ca43705d90f7a1/test/distributed/checkpoint/test_state_dict.py#L146 .

Reply:
If you are comparing 2 identical state_dicts, this should be enough. However, if you need to compare with non-converted optimizer state_dict, a customized function is required as non-converted optimizer state use parameter id.